tokentrim/tokentrim.py

```diff
@@ -80,22 +88,27 @@ def shorten_message_to_fit_limit(message: Dict[str, Any], tokens_needed: int,
   """
   Shorten a message to fit within a token limit by removing characters from the middle.
   """
+  iterations = 0
+
+  encoding = get_encoding(model)
 
   content = message["content"]
 
   while True:
+    iterations += 1
+
     total_tokens = num_tokens_from_messages([message], model)
 
-    if total_tokens <= tokens_needed:
+    if total_tokens <= tokens_needed or iterations >= MAX_ITERATIONS:
       break
 
     ratio = (tokens_needed) / total_tokens
 
-    new_length = int(len(content) * ratio)
+    new_length = int(len(encoding.encode(content)) * ratio)
 
     half_length = new_length // 2
-    left_half = content[:half_length]
-    right_half = content[-half_length:]
+    left_half = encoding.decode(encoding.encode(content[:half_length]))
+    right_half = encoding.decode(encoding.encode(content[-half_length:]))
```

Comment:
Now we're using the actual tokens rather than characters.

Reply:
I can't believe this worked without that.